### ISSUE TITLE
Solution of making S2E compatible with symbol related XMM instructions. 

### ISSUE THREAD

**Issue by moliam** [S2E/libcpu#5]
I want to use s2e for symbolically executing the XMM instructions, such as ucomiss,  cvtdq2pd, etc. However, the current version of S2E seems incompatible with this situation very well because all these instructions are softly handled with helper functions which are compiled with gcc into binary form. That means if these instruction are operating on some symbols, symbolic propagation information may be lost. 
I am wondering that since op_helper.c can be compiled with LLVM compiler and integrated into the symbolic engine to handle some symbolic operation, can we compile these XMM helper function similarly ? ( at the same time XMM reg in CPUArchState need to be move ahead of EIP for making it a symbolic cpu area).   Can this idea solve the problem? Thank you very much.

**Comment by vitalych** [comment]
op_helper.bc is already compiled to LLVM. The XMM functions are in the other files, which are handled externally. In order to fix this, you need to do the following:

- Compile all XMM helpers to LLVM by tweaking the makefiles
- Move XMM registers to symbolic region of the CPUState (i.e., before EIP)
- Wrap all accesses to XMM registers (similarly to what's done for general purpose registers).
- There may be  tweaks required in tcg-llvm.cpp to accomodate for extra symbolic registers.

**Comment by adrianherrera** [comment]
There's also a discussion of this issue here: https://github.com/S2E/libs2eplugins/issues/14

**Comment by moliam** [comment]
Thank you very much. After some changes to the makefile and source code,  I have succeeded in handling the XMM helper funcs in LLVM mode. However, I found that it won't be very useful unless the helper functions can be implemented in a summary way. Because there are many branch operation in the helper functions,  keeping one execution state away from the whole view of the helper function. 
For example, for the following instruction:
```
ucomisd     xmm0,[ ADDR1 ]
jnz ADDR2
```
At instruction "jnz",  the oppsite solution may be unsolvable because the LLVM implementation code for ucomisd contains many branches and make one execution state only walk through part of the operation ucomisd. That means the constraints information collected at jnz is not complete.   As a solution, ucomisd should be summarized into a whole formula. 
I don't know whether I make myself clearly understood......
